rules/review-microservice-boundaries/rule.md

```diff
@@ -0,0 +1,52 @@
+---
+type: rule
+title: Do you regularly review your Microservice boundaries?
+uri: review-microservice-boundaries
+authors:
+  - title: Luke Cook
+    url: https://www.ssw.com.au/people/luke-cook
+related:
+
+created: 2024-10-25T01:09:23.840Z
+guid: f10dba3f-e56a-4289-9127-9cb48e40a7f6
+archivedreason: null
+---
+Microservice architecture poses some unique challenges that can often trip up even the most experienced development teams. One of the hairiest problems is creating well-defined boundaries between your services, and the ideal Microservice architecture often differs from what was originally drawn on the whiteboard. So how do you tell if you got it right?
+<!-- endintro -->
+
+## Map your Microservice architecture
+
+* Draw your Microservices on a whiteboard
+* Whenever one Microservice needs to communicate with another, draw a line between them. If there are several distinct integration points between the same 2 Microservices, add a "+1" or "+2" or "+N" to the line, indicating how many distinct calls those services share. For every paired Microservice, list all the integration points, with particular focus on *why* this integration exists, and the *frequency* in which it occurs (ie rarely, frequently, every time). For example: `OrdersService.Create(...)->Customers.GetCustomer(...). Frequency: every request.`.
+
+![Microservice maps can get intense](bad-example-microservice-boundaries.png)
+**Figure: Incorrect boundaries result in spaghetti services**
+
+### Identify hotspots
+
+* Flag "chatty" services - ie, those services that have the most number of integration points
+* Flag *blocking calls* - ie, if a particular Microservice must wait on downstream/dependent service call before completing
+* Flag *local caches* - ie, if Microservice A keeps a local cache of Microservice B's data
+
+Hotspots are often a **code smell** that could point to Microservices being tightly coupled. These are telltale signs that the boundaries between those services may be in need of adjusting.
+
+## Compare and contrast with BAU
+
+In order to measure your Microservice boundary health, you should see what path a "typical" use case takes through your Microservice.
```

Comment:
Another option here is 'Context Mapping', where by you look at the relationships between different bounded contexts.

Reply:
Done